src/install/install.zig

```diff
@@ -2788,12 +2788,31 @@ pub const PackageManager = struct {
                 switch (index) {
                     .PackageID => |existing_id| {
                         if (existing_id < resolutions.len) {
+                            const existing_resolution = resolutions[existing_id];
+                            if (this.resolutionSatisfiesDependency(existing_resolution, version)) {
+                                successFn(this, dependency_id, existing_id);
+                                return .{
+                                    .package = this.lockfile.packages.get(existing_id),
+                                };
+                            }
+
                             const res_tag = resolutions[existing_id].tag;
                             const ver_tag = version.tag;
                             if ((res_tag == .npm and ver_tag == .npm) or (res_tag == .git and ver_tag == .git) or (res_tag == .github and ver_tag == .github)) {
+                                const existing_package = this.lockfile.packages.get(existing_id);
+                                this.log.addWarningFmt(
+                                    null,
+                                    logger.Loc.Empty,
+                                    this.allocator,
+                                    "incorrect peer dependency \"{}@{}\"",
+                                    .{
+                                        existing_package.name.fmt(this.lockfile.buffers.string_bytes.items),
+                                        existing_package.resolution.fmt(this.lockfile.buffers.string_bytes.items),
+                                    },
+                                ) catch unreachable;
                                 successFn(this, dependency_id, existing_id);
                                 return .{
-                                    .package = this.lockfile.packages.get(existing_id),
+                                    .package = existing_package,
```

Comment:
this.lockfile.packages.get(existing_id)
```suggestion
                                     // we must fetch it from the packages array again, incase the package array mutates the value in the `successFn`
                                    .package = this.lockfile.packages.get(existing_id),
```